File: packages/llm-test-harness/src/cassette/scrubbers.ts
import type { CassetteInteraction, CassetteMessage } from '../types.js'

type Scrubber = (interaction: CassetteInteraction) => CassetteInteraction

const REDACTED = '[REDACTED]'

// Patterns for common secrets
const PATTERNS: Record<string, RegExp> = {
  anthropicKey: /sk-ant-[A-Za-z0-9\-_]{20,}/g,
  openaiKey: /sk-[A-Za-z0-9]{20,}/g,
  bearerToken: /Bearer\s+[A-Za-z0-9\-_\.]{20,}/gi,
  email: /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g,
  uuid: /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/gi,
  ipv4: /\b(?:\d{1,3}\.){3}\d{1,3}\b/g,
}

function scrubString(value: string, pattern: RegExp, replacement: string): string {
  return value.replace(pattern, replacement)
}

function scrubMessages(messages: CassetteMessage[], pattern: RegExp, replacement: string): CassetteMessage[] {
  return messages.map((m) => ({
    ...m,
    content: typeof m.content === 'string' ? scrubString(m.content, pattern, replacement) : m.content,
  }))
}

function scrubInteraction(interaction: CassetteInteraction, pattern: RegExp, replacement: string): CassetteInteraction {
  return {
    ...interaction,
    request: {
      ...interaction.request,
      system: interaction.request.system
        ? scrubString(interaction.request.system, pattern, replacement)
        : interaction.request.system,
      messages: scrubMessages(interaction.request.messages, pattern, replacement),
    },
    response: {
      ...interaction.response,
      content: interaction.response.content?.map((b) => ({
        ...b,
        text: b.text ? scrubString(b.text, pattern, replacement) : b.text,
      })),
    },
  }
}

/**
 * Scrubbers are composable helpers for use with `onBeforeRecord`.
 *
 * @example
 * const harness = new Harness({
 *   cassettesDir: './cassettes',
 *   mode: 'record',
 *   onBeforeRecord: Scrubbers.standard(),
 * })
 */
export const Scrubbers = {
  /** Redact Anthropic and OpenAI API keys */
  apiKey(): Scrubber {
    return (i) => {
      let out = scrubInteraction(i, PATTERNS.anthropicKey!, REDACTED)
      out = scrubInteraction(out, PATTERNS.openaiKey!, REDACTED)
      out = scrubInteraction(out, PATTERNS.bearerToken!, `Bearer ${REDACTED}`)
      return out
    }
  },

  /** Redact email addresses */
  email(replacement = REDACTED): Scrubber {
    return (i) => scrubInteraction(i, PATTERNS.email!, replacement)
  },

  /** Redact UUIDs */
  uuid(replacement = '00000000-0000-0000-0000-000000000000'): Scrubber {
    return (i) => scrubInteraction(i, PATTERNS.uuid!, replacement)
  },

  /** Redact IPv4 addresses */
  ipAddress(replacement = '0.0.0.0'): Scrubber {
    return (i) => scrubInteraction(i, PATTERNS.ipv4!, replacement)
  },

  /** Redact any custom pattern */
  custom(pattern: RegExp, replacement = REDACTED): Scrubber {
    return (i) => scrubInteraction(i, pattern, replacement)
  },

  /**
   * Combine multiple scrubbers into one — applied left to right.
   * @example
   * Scrubbers.combine(Scrubbers.apiKey(), Scrubbers.email())
   */
  combine(...scrubbers: Scrubber[]): Scrubber {
    return (i) => scrubbers.reduce((acc, s) => s(acc), i)
  },

  /**
   * Standard preset: redacts API keys, bearer tokens, emails, UUIDs.
   * Good starting point for most projects.
   */
  standard(): Scrubber {
    return Scrubbers.combine(
      Scrubbers.apiKey(),
      Scrubbers.email(),
      Scrubbers.uuid(),
    )
  },
}

File: packages/llm-test-harness-python/llm_test_harness/cassette/scrubbers.py
"""
Scrubbers — composable helpers for sanitising cassettes before they are written.

Use with the ``on_before_record`` hook::

    harness = Harness(
        cassettes_dir="./cassettes",
        mode="record",
        on_before_record=Scrubbers.standard(),
    )
"""
from __future__ import annotations

import re
from typing import Callable

from ..types import CassetteInteraction, CassetteMessage

Scrubber = Callable[[CassetteInteraction], CassetteInteraction]
REDACTED = "[REDACTED]"

_PATTERNS: dict[str, re.Pattern[str]] = {
    "anthropic_key": re.compile(r"sk-ant-[A-Za-z0-9\-_]{20,}"),
    "openai_key": re.compile(r"sk-[A-Za-z0-9]{20,}"),
    "bearer_token": re.compile(r"Bearer\s+[A-Za-z0-9\-_.]{20,}", re.IGNORECASE),
    "email": re.compile(r"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}"),
    "uuid": re.compile(r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}", re.IGNORECASE),
    "ipv4": re.compile(r"\b(?:\d{1,3}\.){3}\d{1,3}\b"),
}


def _scrub_str(value: str, pattern: re.Pattern[str], replacement: str) -> str:
    return pattern.sub(replacement, value)


def _scrub_messages(messages: list[CassetteMessage], pattern: re.Pattern[str], replacement: str) -> list[CassetteMessage]:
    from dataclasses import replace as dc_replace
    return [dc_replace(m, content=_scrub_str(m.content, pattern, replacement)) for m in messages]


def _scrub_interaction(interaction: CassetteInteraction, pattern: re.Pattern[str], replacement: str) -> CassetteInteraction:
    from dataclasses import replace as dc_replace
    new_system = (
        _scrub_str(interaction.request.system, pattern, replacement)
        if interaction.request.system
        else interaction.request.system
    )
    new_messages = _scrub_messages(interaction.request.messages, pattern, replacement)
    new_content = None
    if interaction.response.content:
        from dataclasses import replace as r
        new_content = [
            r(b, text=_scrub_str(b.text, pattern, replacement)) if b.text else b
            for b in interaction.response.content
        ]
    new_req = dc_replace(interaction.request, system=new_system, messages=new_messages)
    new_resp = dc_replace(interaction.response, content=new_content)
    return dc_replace(interaction, request=new_req, response=new_resp)


class Scrubbers:
    """Composable scrubber factories. All methods return a ``Scrubber`` callable."""

    @staticmethod
    def api_key() -> Scrubber:
        """Redact Anthropic and OpenAI API keys and Bearer tokens."""
        def _scrub(i: CassetteInteraction) -> CassetteInteraction:
            i = _scrub_interaction(i, _PATTERNS["anthropic_key"], REDACTED)
            i = _scrub_interaction(i, _PATTERNS["openai_key"], REDACTED)
            i = _scrub_interaction(i, _PATTERNS["bearer_token"], f"Bearer {REDACTED}")
            return i
        return _scrub

    @staticmethod
    def email(replacement: str = REDACTED) -> Scrubber:
        """Redact email addresses."""
        return lambda i: _scrub_interaction(i, _PATTERNS["email"], replacement)

    @staticmethod
    def uuid(replacement: str = "00000000-0000-0000-0000-000000000000") -> Scrubber:
        """Redact UUIDs."""
        return lambda i: _scrub_interaction(i, _PATTERNS["uuid"], replacement)

    @staticmethod
    def ip_address(replacement: str = "0.0.0.0") -> Scrubber:
        """Redact IPv4 addresses."""
        return lambda i: _scrub_interaction(i, _PATTERNS["ipv4"], replacement)

    @staticmethod
    def custom(pattern: re.Pattern[str] | str, replacement: str = REDACTED) -> Scrubber:
        """Redact any custom regex pattern."""
        pat = re.compile(pattern) if isinstance(pattern, str) else pattern
        return lambda i: _scrub_interaction(i, pat, replacement)

    @staticmethod
    def combine(*scrubbers: Scrubber) -> Scrubber:
        """Combine multiple scrubbers, applied left to right."""
        def _combined(i: CassetteInteraction) -> CassetteInteraction:
            for s in scrubbers:
                i = s(i)
            return i
        return _combined

    @staticmethod
    def standard() -> Scrubber:
        """Preset: redacts API keys, bearer tokens, emails, and UUIDs."""
        return Scrubbers.combine(
            Scrubbers.api_key(),
            Scrubbers.email(),
            Scrubbers.uuid(),
        )
